percona/controllers/pgc/update.go

```diff
@@ -0,0 +1,200 @@
+package pgc
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/robfig/cron/v3"
+	log "github.com/sirupsen/logrus"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/percona/percona-postgresql-operator/internal/kubeapi"
+	"github.com/percona/percona-postgresql-operator/percona/controllers/pgcluster"
+	"github.com/percona/percona-postgresql-operator/percona/controllers/pgreplica"
+	"github.com/percona/percona-postgresql-operator/percona/controllers/version"
+	crv1 "github.com/percona/percona-postgresql-operator/pkg/apis/crunchydata.com/v1"
+)
+
+func (c *Controller) updateVersion(oldCluster, newCluster *crv1.PerconaPGCluster) error {
+	if oldCluster.Spec.Backup.Image == newCluster.Spec.Backup.Image &&
+		oldCluster.Spec.Backup.BackrestRepoImage == newCluster.Spec.Backup.BackrestRepoImage &&
+		oldCluster.Spec.PGBadger.Image == newCluster.Spec.PGBadger.Image &&
+		oldCluster.Spec.PGBouncer.Image == newCluster.Spec.PGBouncer.Image &&
+		oldCluster.Spec.PGPrimary.Image == newCluster.Spec.PGPrimary.Image {
+		return nil
+	}
+
+	return smartUpdateCluster(c.Client, newCluster, oldCluster)
+}
+
+func (c *Controller) deleteEnsureVersion(jobName string) {
+	c.crons.crons.Remove(cron.EntryID(c.crons.ensureVersionJobs[jobName].ID))
+	delete(c.crons.ensureVersionJobs, jobName)
+}
+
+func (c *Controller) scheduleUpdate(newCluster *crv1.PerconaPGCluster) error {
+	jn := newCluster.Name + "/ensureVersion"
+	schedule, ok := c.crons.ensureVersionJobs[jn]
+	if newCluster.Spec.UpgradeOptions == nil {
+		if ok {
+			c.deleteEnsureVersion(jn)
+		}
+		return nil
+	}
+	if ok && schedule.CronSchedule == newCluster.Spec.UpgradeOptions.Schedule {
+		return nil
+	}
+	if ok {
+		c.deleteEnsureVersion(jn)
+	}
+	id, err := c.crons.crons.AddFunc(newCluster.Spec.UpgradeOptions.Schedule, func() {
+		log.Println("Smart update started")
+		err := version.EnsureVersion(newCluster, version.VersionServiceClient{
+			OpVersion: newCluster.ObjectMeta.Labels["pgo-version"],
+		})
+		if err != nil {
+			log.Errorf("schedule update: ensure version %s", err)
+		}
+		oldCluster, err := c.Client.CrunchydataV1().PerconaPGClusters(newCluster.Namespace).Get(context.TODO(), newCluster.Name, metav1.GetOptions{})
+		if err != nil {
+			log.Errorf("update perconapgcluster: %s", err)
```

Comment:
error message should be something like `get perconapgcluster`

Reply:
https://github.com/percona/percona-postgresql-operator/pull/152/commits/d58e7ffba38850624ce2d100a891ced7f50a7cd8